### PR TITLE
Bs 41

### LINKED ISSUE #41: Подключить сервис бюджета к UI

### PR DESCRIPTION
**Pull request by antonhoang**
## Details 
- connect budget service to UI 
- add some tasks in doc `BudgetUseCase.md `for discuss

## Affected 
- [x] Modules/BudgetDetails

## Screenshots
- i think we can put somewhere the period days for user. Need your advice and ideas

Screenshot 2022-07-17 at 16 20 18

### CODE REVIEW

File: BudgetServiceTests/BudgetUseCase.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+
+# Обновления бюджета 
+- Пример создан бюджет: на 10000 $ на 30 дней
+
+- Мы знаем что считать нужно количество денег выделено на бюджет равно 10000 $, но считать сколько можно потратить в день только из тех денег которых осталось потому что пользователь может создавать новые транзакции  
+
+- к примеру у юзера есть 10000 $ на 30 дней, если они ничего не ввел то в день он может тратить 10000 / 30 = 333.33 $
+
+- если он все-таки ввел новую транзакцию на наркотики 1000 $, у него осталось 9000$
```
Comment (Valter4578) [inline BudgetServiceTests/BudgetUseCase.md:37]: 😳😳😳

Comment (Valter4578) [inline BudgetServiceTests/BudgetUseCase.md:37]: 😳😳😳 
Todo: добавить стандартную категорию - наркотики

File: bonsai/Modules/BudgetDetails/BudgetViewModel.swift [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+   
+   init(budgetService: BudgetServiceProtocol, transactions: [Transaction]) {
+      self.budgetService = budgetService
+      self.transactions = transactions
+      /*
+       for manual test
+       1 delete app from simulator and uncomment createBudget()
+       2 create some transactions
+       */
+//      budgetService.createBudget(name: "Tailand", with: 9000, on: 30)
```
Comment (Valter4578) [inline bonsai/Modules/BudgetDetails/BudgetViewModel.swift]: Why does name not shown on the screen?

Simulator Screen Shot - iPhone 13 - 2022-07-26 at 13 50 42

**Review by Valter4578 — COMMENTED** [review]
https://user-images.githubusercontent.com/38043698/180943149-24f4b078-3077-4d88-896e-78cc6cf4c57a.mov
Income transactions counts in "money spent" section in budget screen

**Review by Valter4578 — COMMENTED** [review]
Simulator Screen Shot - iPhone 13 - 2022-07-26 at 13 54 58
Transaction doesn't appear after creating in transaction list. You should do it within this task

**Review by Valter4578 — CHANGES_REQUESTED** [review]
Needs additional improvements

File: bonsai/Modules/BudgetDetails/BudgetViewModel.swift (hunk 2 above)
Comment (antonhoang) [reply bonsai/Modules/BudgetDetails/BudgetViewModel.swift]: updated+

**Review by Valter4578 — APPROVED** [review]

File: bonsai/Modules/BudgetDetails/BudgetViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+   init(budgetService: BudgetServiceProtocol, transactions: [Transaction]) {
+      self.budgetService = budgetService
+      let nonIncomeTransactions = transactions.filter { $0.type != .income }
+      self.transactions = nonIncomeTransactions
+      /*
+       for manual test
+       1 delete app from simulator and uncomment createBudget()
+       2 create some transactions
+       */
+      //      budgetService.createBudget(name: "Tailand", with: 9000, on: 30)
```
Comment (VladimirBrejcha) [inline bonsai/Modules/BudgetDetails/BudgetViewModel.swift]: Old comment? Probably not needed

File: bonsai/Modules/BudgetDetails/BudgetViewModelAssembler.swift [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+struct BudgetViewModelAssembler: Assembler {
+   private var transactions: [Transaction] = []
+
+   init(mainContext: NSManagedObjectContext = DataController.sharedInstance.container.viewContext) {
+      let fetchRequest: NSFetchRequest<Transaction> = Transaction.fetchRequest()
+      do {
+         let transactions = try mainContext.fetch(fetchRequest)
+         self.transactions = transactions
+      } catch {
+
```
Comment (VladimirBrejcha) [inline bonsai/Modules/BudgetDetails/BudgetViewModelAssembler.swift]: I suggest to at least add assert in catch block

File: bonsai/Assembler/Assembler.swift [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+//  bonsai
+//
+//  Created by antuan.khoanh on 17/07/2022.
+//
+
+import Foundation
+
+protocol Assembler {
+   associatedtype T
+   func assembly() -> T
```
Comment (VladimirBrejcha) [inline bonsai/Assembler/Assembler.swift]: Why do we need this protocol?

File: bonsai/Modules/BudgetDetails/BudgetViewModel.swift (hunk 3 above)
Comment (antonhoang) [reply bonsai/Modules/BudgetDetails/BudgetViewModel.swift]: deleted +

File: bonsai/Modules/BudgetDetails/BudgetViewModelAssembler.swift (hunk 4 above)
Comment (antonhoang) [reply bonsai/Modules/BudgetDetails/BudgetViewModelAssembler.swift]: updated +

File: bonsai/Assembler/Assembler.swift (hunk 5 above)
Comment (antonhoang) [reply bonsai/Assembler/Assembler.swift]: deleted +

**Pull request merged**